File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void solve(string brackets, int open, int close, vector<string>& res){
        if(open == 0 and close == 0){
            res.push_back(brackets);
            return;
        }
        if(open == close){
            brackets.push_back('(');
            open--;
            solve(brackets, open, close, res);
        }
        else if(open == 0){
            brackets.push_back(')');
            close--;
            solve(brackets, open, close, res);
        }else{
            solve(brackets+'(', open-1,close,res);
            solve(brackets+')', open,close-1, res);
        }
    }
    vector<string> generateParenthesis(int n) {
        int open = n;
        int close = n;
        vector<string>res;
        solve("", open, close,res);
        return res;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def solve(self,ans,o,c,n,s):
        if o+c>2*n:
            return
        
        if len(s) == n*2 and o==c:
            ans.append(s)
            print(s)
            return
        
        if c < o:
            self.solve(ans,o,c+1,n,s+")")
        self.solve(ans,o+1,c,n,s+"(")
        
        return
    
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.solve(ans,0,0,n,"")
        return ans